task/thermal/src/main.rs

```diff
@@ -175,38 +185,33 @@ impl<'a, B: BspT> idl::InOrderThermalImpl for ServerImpl<'a, B> {
             30 => I2cWatchdog::ThirtySeconds,
             _ => return Err(ThermalError::InvalidWatchdogTime.into()),
         };
-        ServerImpl::<B>::set_watchdog(self, wd).map_err(Into::into)
+        ServerImpl::set_watchdog(self, wd).map_err(Into::into)
     }
 }
 
-impl<'a, B: BspT> NotificationHandler for ServerImpl<'a, B> {
+impl<'a> NotificationHandler for ServerImpl<'a> {
     fn current_notification_mask(&self) -> u32 {
         TIMER_MASK
     }
 
     fn handle_notification(&mut self, _bits: u32) {
-        self.deadline += TIMER_INTERVAL;
-        self.counter += 1;
-        sys_set_timer(Some(self.deadline), TIMER_MASK);
-
-        match self.mode {
-            ThermalMode::Auto => {
-                if self.counter % CONTROL_RATE == 0 {
-                    // TODO: what to do with errors here?
-                    let _ = self.control.run_control();
-                } else {
-                    let _ = self.control.read_sensors();
+        let now = sys_get_timer().now;
+        if self.deadline >= now {
+            match self.mode {
+                ThermalMode::Auto => {
+                    self.control.run_control(now);
+                }
+                ThermalMode::Manual => {
+                    // Read sensors and post them to the `sensors` task
+                    self.control.read_sensors(now);
+                }
+                ThermalMode::Off => {
+                    panic!("Mode must not be 'Off' when server is running")
```

Comment:
How would this panic affect the fans on the running system?
Are things ever going to be out of sync so that this condition needs to be tolerated? (e.g. engineer manually sets power state, thermal loop restarts due to panic?)

Reply:
I think this should be safe: `ThermalMode::Off` is only used during the brief period between constructing the `struct` and calling `set_mode_manual` on startup.

We only expose APIs to set the mode to manual or auto, so no one should be able to set it to `ThermalMode::Off` during operation, even during unusual debug circumstances.